### PR TITLE
Fix export move_to_device_pass device rewrites

### PR DESCRIPTION
**Pull request by jansel**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.16.0) (oldest at bottom):
* __->__ #185105

`move_to_device_pass` rewrites device-bearing FX nodes and tensor metadata, but it did not recompile the affected GraphModules. Generated Python could therefore retain stale `_assert_tensor_metadata` device arguments and fail when the moved graph ran directly through `ep.graph_module`.

Recompile every GraphModule touched by the pass so its callable code matches the rewritten graph. Export also supports interpreter-only graphs whose in-memory custom objects cannot be emitted as Python, so keep that existing fallback by ignoring only the same `SyntaxError` raised during code generation.

An earlier version tried to preserve parameter identity with `.data`. Freshly exported state aliases the source module, so that moved source parameters while replacing buffers and left the source module mixed-device. It also corrupted traceable wrapper subclasses by moving only the outer TensorImpl. Retain the existing out-of-place state movement instead, with regressions covering source-module reuse and wrapper-subclass children.

I considered overriding `GraphModule.to()` so it rewrites graph IR, but exported graphs are device-specific and `move_to_device_pass` is the supported graph-casting path. Keeping the fix in that pass avoids changing general FX `GraphModule.to()` semantics.

Fixes #163323
Generated by my agent

Test Plan:
```bash
python test/export/test_passes.py TestPasses.test_move_device_to TestPasses.test_move_device_submod TestPasses.test_move_to_device_pass_recompiles_graph_module TestPasses.test_move_to_device_pass_does_not_mutate_source_module TestPasses.test_move_to_device_pass_wrapper_subclass_parameter TestPasses.test_move_to_device_pass_interpreter_only_graph TestPasses.test_move_to_device_pass TestPasses.test_move_device_example_inputs
```

```bash
python test/export/test_export.py TestOneOffModelExportResult.test_assert_tensor_metadata_device_index
```

```bash
lintrunner -a
```

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

File: torch/export/passes/__init__.py [hunk 1]
```diff
@@ -37,20 +37,43 @@ def _get_new_device(
...
             return str(location)
 
+    def _move_tensor(tensor: torch.Tensor) -> torch.Tensor:
+        return tensor.to(_get_new_device(tensor.device, location))
+
+    def _move_parameter(param: torch.nn.Parameter) -> torch.nn.Parameter:
+        with torch.no_grad():
+            param_applied = _move_tensor(param)
+        if torch._has_compatible_shallow_copy_type(param, param_applied):
+            param.data = param_applied
```
Comment (bobrenjc93) [inline torch/export/passes/__init__.py]: For freshly exported programs, `ep.state_dict` contains the source module's actual Parameter objects. Assigning `param.data` therefore moves the source module's parameters, while buffers below are only replaced in the EP dictionary, leaving a retained source module mixed-device and unusable. Either detach EP state before this mutation or establish consistent, documented alias semantics for all state, with a source-module reuse test.

File: torch/export/passes/__init__.py [hunk 2]
```diff
@@ -37,20 +37,43 @@ def _get_new_device(
         else:
             return str(location)
 
+    def _move_tensor(tensor: torch.Tensor) -> torch.Tensor:
+        return tensor.to(_get_new_device(tensor.device, location))
+
+    def _move_parameter(param: torch.nn.Parameter) -> torch.nn.Parameter:
+        with torch.no_grad():
+            param_applied = _move_tensor(param)
+        if torch._has_compatible_shallow_copy_type(param, param_applied):
```
Comment (bobrenjc93) [inline torch/export/passes/__init__.py]: This `.data` path is unsafe for traceable wrapper-subclass parameters. Export supports parameters such as `TwoTensor`, whose child tensors live in Python attributes; `.data` updates only the outer TensorImpl, leaving those children on the original device. Mirror `Module._apply`'s `is_traceable_wrapper_subclass`/`swap_tensors` path and add a subclass-parameter regression test.

File: torch/export/passes/__init__.py [hunk 3]
```diff
@@ -92,6 +115,7 @@ def _get_new_device(
                     else v,
                     node.meta.get("val"),
                 )
+            m.recompile()
```
Comment (bobrenjc93) [inline torch/export/passes/__init__.py]: `ExportedProgram` deliberately catches `SyntaxError` during GraphModule code generation and retains the graph for interpreter execution when custom objects cannot be emitted as Python. This unconditional `recompile()` retries that unsupported generation without a catch, so moving such a valid interpreter-only EP now raises. Preserve the fallback and test an EP containing an in-memory custom-object constant.

Comment (jansel) [reply torch/export/passes/__init__.py]: My agent says

> Addressed by catching only SyntaxError from GraphModule.recompile(), matching the existing ExportedProgram interpreter-only fallback. The new regression injects an in-memory custom object, rewrites device-bearing IR from CPU to meta, and verifies execution through torch.fx.Interpreter.

File: torch/export/passes/__init__.py (hunk 2 above)
Comment (jansel) [reply torch/export/passes/__init__.py]: My agent says

> Addressed by removing the unsafe in-place path. Wrapper-subclass parameters now use the existing out-of-place Tensor.to() path, which moves both the outer wrapper and its child tensors. The new TwoTensor regression verifies that the moved exported state is CUDA-valid while the retained source state remains CPU-valid.

File: torch/export/passes/__init__.py (hunk 1 above)
Comment (jansel) [reply torch/export/passes/__init__.py]: My agent says

> Addressed by removing the in-place parameter-data mutation and retaining the existing out-of-place state replacement. The new regression verifies that both parameters and buffers in the retained source module stay on CPU and remain usable after the exported program moves to CUDA.

File: torch/export/passes/__init__.py [hunk 4]
```diff
@@ -92,6 +92,12 @@ def _get_new_device(
                     else v,
                     node.meta.get("val"),
                 )
+            try:
+                m.recompile()
```
Comment (mlazos) [inline torch/export/passes/__init__.py:96]: _Reviewed by me, this may be an issue._

_AI-generated analysis (Codex):_

> **Nested GraphModules are recompiled repeatedly**
> 
> GraphModule.recompile() recursively recompiles GraphModule children, while modules() subsequently visits those descendants again. An N-deep hierarchy therefore triggers N(N+1)/2 code generations and forward replacements, adding quadratic overhead to this pass.

**Review by mlazos — APPROVED** [review]
